### PR TITLE
bugfix: reduce race condition between get/incr(key)

### PR DESCRIPTION
**Pull request by spacewander**
There may be a potential race condition between `get` and `incr` operation.

Imagine such a scene:
The max connections per ip is 10, and `127.0.0.1` has already 9. There are 4 Nginx workers.
Each worker gets the connections of `127.0.0.1` one by one. They see the current connections is 9 only, so it's safe to increase one to it.
Finally there will be 13 connections($max_conn + $worker_num - 1) for `127.0.0.1`.

By using `incr` first, we could avoid those overcommits.

### CODE REVIEW

File: lib/resty/limit/conn.lua [hunk 1]
```diff
@@ -47,58 +47,38 @@ function _M.incoming(self, key, commit)
...
-                if not conn then
-                    return nil, err
-                end
-
-                self.committed = true
-            else
-                conn = conn + 1
-            end
+        if conn > max + self.burst then
+            dict:incr(key, -1, conn)
```
Comment (agentzh) [inline lib/resty/limit/conn.lua]: No error handling here?

File: lib/resty/limit/conn.lua [hunk 2]
```diff
@@ -47,58 +47,38 @@ function _M.incoming(self, key, commit)
...
         end
 
-        if commit then
-            -- FIXME: we should use incr_or_init here.
-            local new_conn, err = dict:incr(key, 1)
-            if not new_conn then
-                return nil, err
+        -- we return a 0 delay by default
+        return 0, conn
+    else
```
Comment (agentzh) [inline lib/resty/limit/conn.lua]: We can omit this `else` here since the previous branch always returns and never falls through.

**Comment by spacewander** [comment]
@agentzh 
Changes updated.

**Comment by agentzh** [comment]
@spacewander Thanks!

@thibaultcha Mind to have a look at this? Thanks!

File: lib/resty/limit/conn.lua [hunk 3]
```diff
@@ -47,58 +47,41 @@ function _M.incoming(self, key, commit)
...
-            end
+        -- we return a 0 delay by default
+        return 0, conn
+    end
 
-            self.committed = true
-            return 0, new_conn
+    local conn = dict:get(key) or 0
+    if conn >= max then
+        if conn >= max + self.burst then
```
Comment (thibaultcha) [inline lib/resty/limit/conn.lua]: I was about to say, I think we should also nest this case under the `conn > max` branch up there for the `commited` branch, and then I see it is being done here for the non-commited. Should we? It used to be.

**Comment by spacewander** [comment]
@thibaultcha 
Now both `commited` and `non-commited` branches share the same `conn > max` case.
Is it something you want?

**Comment by thibaultcha** [comment]
I think you've updated the patch since yesterday. It wasn't exactly what I meant but no big deal, I think this approach is good as well, and reduces cognitive load :+1:

**Comment by spacewander** [comment]
@agentzh 
Are there further changes is in need?

**Comment by agentzh** [comment]
@spacewander It looks good to me. Waiting on @doujiang24's feedback. Sorry for the delay!

**Comment by agentzh** [comment]
@spacewander Merged. Thanks!

**Pull request closed without merging**